src/ssl.cpp

```diff
@@ -193,57 +193,57 @@ int CGLS(const AlgIn& data, const double lambda, const int cgitermax,
   return optimality;
 }
 
-int L2_SVM_MFN(const AlgIn& data, struct options* Options,
-               struct vector_double* Weights,
-               struct vector_double* Outputs, double cpos, double cneg) {
+int L2_SVM_MFN(const AlgIn& data, options* Options,
+               vector_double& Weights,
+               vector_double& Outputs, double cpos, double cneg) {
   /* Disassemble the structures */
   Timer tictoc;
   double** set = data.vals;
   const double* Y = data.Y;
-  int n = Weights->d;
+  int n = Weights.d;
   const int m = data.m;
   double lambda = Options->lambda;
   double epsilon = BIG_EPSILON;
   int cgitermax = SMALL_CGITERMAX;
-  double* w = Weights->vec;
-  double* o = Outputs->vec;
+  double* w = Weights.vec;
+  double* o = Outputs.vec;
   double F_old = 0.0;
   double F = 0.0;
   double diff = 0.0;
   int ini = 0;
   int n0 = n-1;
   int inc = 1;
-  vector_int* ActiveSubset = new vector_int[1];
-  ActiveSubset->vec = new int[m];
-  ActiveSubset->d = m;
+  vector_int ActiveSubset;
+  ActiveSubset.vec = new int[m];
+  ActiveSubset.d = m;
   // initialize
   F = 0.5 * lambda * ddot_(&n, w, &inc, w, &inc);
   int active = 0;
   int inactive = m - 1; // l-1
   for (int i = 0; i < m; i++) {
     diff = 1 - Y[i] * o[i];
     if (diff > 0) {
-      ActiveSubset->vec[active] = i;
+      ActiveSubset.vec[active] = i;
       active++;
       // C[i]
       F += 0.5 * ((Y[i]==1)? cpos : cneg) * diff * diff;
     } else {
-      ActiveSubset->vec[inactive] = i;
+      ActiveSubset.vec[inactive] = i;
       inactive--;
     }
   }
-  ActiveSubset->d = active;
+  ActiveSubset.d = active;
   int iter = 0;
   int opt = 0;
   int opt2 = 0;
-  vector_double* Weights_bar = new vector_double[1];
-  vector_double* Outputs_bar = new vector_double[1];
+  vector_double Weights_bar;
+  vector_double Outputs_bar;
   double* w_bar = new double[n];
   double* o_bar = new double[m];
```

Comment:
Please note that I did not write this code, I only modified what was already there to make it look cleaner.

In practice the new statement is called on members of class objects of type vector_double instead of calling new on the entire class object. This way, memory handling is done internally by the object itself in its destructor. Later in this same piece of code w_bar and o_bar are assigned as variables to a vector_double-object if I recall correctly. Is this still ok?

Reply:
Yes thanks, I missed that.